lib/Dialect/FIRRTL/Import/FIRParser.cpp

```diff
@@ -3506,24 +3531,66 @@ FIRCircuitParser::parseCircuit(const llvm::MemoryBuffer *annotationsBuf) {
 
 DoneParsing:
 
+  auto reduceTargets =
+      [](const SmallVector<StringAttr> &acc,
+         SmallVector<StringAttr> x) -> SmallVector<StringAttr> {
+    if (x.empty())
+      return acc;
+    // Collisions are impossible, just append.
+    SmallVector<StringAttr> retval(acc);
+    retval.append(x.begin(), x.end());
+    return retval;
+  };
+
   // Now that we've parsed all the prototypes and created all the module ops,
   // go ahead and parse all their bodies.  This can be done in parallel.
   if (getContext()->isMultithreadingEnabled()) {
     mlir::ParallelDiagnosticHandler diagHandler(getContext());
     std::atomic<bool> anyFailed{false};
-    llvm::parallelForEachN(0, deferredModules.size(), [&](size_t index) {
-      diagHandler.setOrderIDForThread(index);
-      if (parseModuleBody(deferredModules[index]))
-        anyFailed = true;
-    });
+    std::atomic<unsigned> threadID{0};
+    auto newTargets = llvm::parallelTransformReduce(
+        deferredModules.begin(), deferredModules.end(),
+        SmallVector<StringAttr>(), reduceTargets, [&](auto module) {
+          diagHandler.setOrderIDForThread(threadID++);
+          if (parseModuleBody(module))
+            anyFailed = true;
+          return module.targets;
+        });
+    targets.append(newTargets.begin(), newTargets.end());
     if (anyFailed)
       return failure();
   } else {
     for (DeferredModuleToParse &deferredModule : deferredModules) {
       if (parseModuleBody(deferredModule))
         return failure();
+      targets = reduceTargets(targets, deferredModule.targets);
+    }
+  }
+
+  // Verify that all annotations map to valid targets.  Error if any do not.
+  llvm::StringSet<> targetsSet;
```

Comment:
BTW, StringSet takes an Allocator argument.  You might see if it is worth using a BumpPointerAllocator (if StringSet survives)

Reply:
`StringSet` did survive and I switched to using a `BumpPointerAllocator`.  Due to the other cases being caught in #1280 and #1327, `StringSet` works.  If, however, we want to unify the "Annotation does not apply" checks to all use the same infrastructure, this would need to change to something like a `SmallSet<StringRef, llvm::BitVector>` to track both which target was found _and which of its Annotations were applied_.  As it stands @hanchenye's approach can give better error messages than this alternative because he can point to a specific circuit location instead of a generic `~Foo|Bar>baz[0]` did not exist.

What would be a good way to test the choice of allocator?  See how the performance scales for the two approaches as annotations increase?  I may need to see if @drom can add a flag to generate large circuits where _everything_ is annotated. :smile: